### PR TITLE
ref: result 1/n (make monitor default to checkpoint_on to simplify re…

### PR DESCRIPTION
**Pull request by williamFalcon**
clearing up result confusion.

- Set monitor to checkpoint_on in ModelCheckpoint
- Add error when restore checkpoint has more epochs than max epochs

### CODE REVIEW

File: docs/source/lightning_module.rst [hunk 1]
```diff
@@ -288,6 +288,8 @@ For cases like production, you might want to iterate different models inside a L
             y_hat = self.model(x)
             loss = F.cross_entropy(y_hat, y)
             acc = FM.accuracy(y_hat, y)
+
+            # loss is a scalar. The Checkpoint Callback is monitoring 'checkpoint_on'
```
Comment (ananthsub) [inline docs/source/lightning_module.rst:292]: the loss is a Tensor
in case people read scalar as raw float value

File: pytorch_lightning/callbacks/early_stopping.py [hunk 2]
```diff
@@ -42,8 +42,7 @@ class EarlyStopping(Callback):
     r"""
 
     Args:
-        monitor: quantity to be monitored. Default: ``'val_loss'``.
-            .. note:: Has no effect when using `EvalResult` or `TrainResult`
+        monitor: quantity to be monitored. Default: ``'early_stop_on'``.
```
Comment (ananthsub) [inline pytorch_lightning/callbacks/early_stopping.py:45]: can the early stopping changes be split out to a separate PR?

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 3]
```diff
@@ -420,6 +402,10 @@ def on_validation_end(self, trainer, pl_module):
             if self.last_model_path and self.last_model_path != filepath:
                 self._del_model(self.last_model_path)
 
+    def _is_valid_monitor_key(self, metrics):
+        is_valid = self.monitor in metrics or len(metrics) == 0
+        return is_valid
```
Comment (ananthsub) [inline pytorch_lightning/callbacks/model_checkpoint.py:407]: can be inlined. what happens if `metrics` is empty? it seems like this is fine if save_top_k <= 0? 
```suggestion
        return self.monitor in metrics or len(metrics) == 0
```

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 4]
```diff
@@ -342,19 +327,16 @@ def on_validation_end(self, trainer, pl_module):
...
+        # validate metric
+        if not self._is_valid_monitor_key(metrics):
+            keys = list(metrics.keys())
+            m = f"""
+                ModelCheckpoint(monitor='{self.monitor}') not found in the returned metrics ({keys}),
+                "did you call result.log(f'{self.monitor}', tensor)?"""
+            raise MisconfigurationException(m)
 
         if self.save_top_k == 0:
             # no models are saved
```
Comment (ananthsub) [inline pytorch_lightning/callbacks/model_checkpoint.py:342]: this check should be to right after line 327 so we can exit earlier

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 5]
```diff
@@ -145,6 +146,14 @@ def restore_training_state(self, checkpoint):
         self.trainer.global_step = checkpoint['global_step']
         self.trainer.current_epoch = checkpoint['epoch']
 
+        # crash if max_epochs is lower than the current epoch from the checkpoint
```
Comment (ananthsub) [inline pytorch_lightning/trainer/connectors/checkpoint_connector.py:149]: good catch! this could be a separate PR since it's independent of the checkpoint_on changes

File: pytorch_lightning/trainer/connectors/logger_connector.py [hunk 6]
```diff
@@ -121,13 +121,23 @@ def _log_on_evaluation_epoch_end_metrics(self, eval_results, using_eval_result):
                         flat = {'val_loss': eval_result}
                     else:
                         flat = flatten_dict(eval_result)
```
Comment (ananthsub) [inline pytorch_lightning/trainer/connectors/logger_connector.py:123]: nit: there's this implementation and also one in the lightning loggers base for flattening dicts. it'd be nice to have one utility so they don't diverge

**Review by ananthsub — COMMENTED** [review]
much cleaner!

File: pytorch_lightning/callbacks/early_stopping.py (hunk 2 above)
Comment (williamFalcon) [reply pytorch_lightning/callbacks/early_stopping.py:45]: yeah, i initially tried but since val_loss was used for both it made it super difficult

File: pytorch_lightning/callbacks/model_checkpoint.py (hunk 4 above)
Comment (williamFalcon) [reply pytorch_lightning/callbacks/model_checkpoint.py:342]: there's no real difference though. You need the metrics dict on 330 to know if there's an error

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py (hunk 5 above)
Comment (williamFalcon) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py:149]: agreed, but the tests to verify the new changes were failing because of this bug... let me see if i can put in a different PR and merge first

File: pytorch_lightning/trainer/connectors/logger_connector.py (hunk 6 above)
Comment (williamFalcon) [reply pytorch_lightning/trainer/connectors/logger_connector.py:123]: makes sense. we'll merge during the continuing refactors

File: pytorch_lightning/callbacks/model_checkpoint.py (hunk 4 above)
Comment (ananthsub) [reply pytorch_lightning/callbacks/model_checkpoint.py:342]: i guess it's not clear on the UI. I meant moving the `if self.save_top_k == 0` check up to even before we check metrics. if `save_top_k = 0`, then there's no need to check metrics since there are no checkpoints to save

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py (hunk 5 above)
Comment (williamFalcon) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py:149]: https://github.com/PyTorchLightning/pytorch-lightning/pull/3580/files

File: pytorch_lightning/callbacks/model_checkpoint.py (hunk 4 above)
Comment (williamFalcon) [reply pytorch_lightning/callbacks/model_checkpoint.py]: done
